File: src/1721_swapping_nodes_in_a_linked_list.js
function swapNodes(head, k) {
    let rightPointer = head;
    for (let i = 1; i < k; i++) rightPointer = rightPointer.next;
    const leftKthNode = rightPointer;
    let leftPointer = head;
    let rightKthNode = head;
    while (rightPointer !== null) {
        rightKthNode = leftPointer;
        rightPointer = rightPointer.next;
        leftPointer = leftPointer.next;
    }
    [leftKthNode.val, rightKthNode.val] = [rightKthNode.val, leftKthNode.val];
    return head;
}

module.exports = { swapNodes };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(listNodeToString(swapNodes(new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5, null))))), 2)), listNodeToString(new ListNode(1, new ListNode(4, new ListNode(3, new ListNode(2, new ListNode(5, null)))))));
    assert.deepStrictEqual(listNodeToString(swapNodes(new ListNode(7, new ListNode(9, new ListNode(6, new ListNode(6, new ListNode(7, new ListNode(8, new ListNode(3, new ListNode(0, new ListNode(9, new ListNode(5, null)))))))))), 5)), listNodeToString(new ListNode(7, new ListNode(9, new ListNode(6, new ListNode(6, new ListNode(8, new ListNode(7, new ListNode(3, new ListNode(0, new ListNode(9, new ListNode(5, null))))))))))));
}

File: src/1721_swapping_nodes_in_a_linked_list.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { val, next: None }
    }
}

struct Solution;

impl Solution {
    pub fn swap_nodes(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
        let mut values = list_to_vec(&head);
        let left = k as usize - 1;
        let right = values.len() - k as usize;
        values.swap(left, right);
        vec_to_list(values)
    }
}

fn list_to_vec(head: &Option<Box<ListNode>>) -> Vec<i32> {
    let mut values = Vec::new();
    let mut current = head.as_ref();
    while let Some(node) = current {
        values.push(node.val);
        current = node.next.as_ref();
    }
    values
}

fn vec_to_list(values: Vec<i32>) -> Option<Box<ListNode>> {
    let mut head = None;
    for value in values.into_iter().rev() {
        let mut node = Box::new(ListNode::new(value));
        node.next = head;
        head = Some(node);
    }
    head
}

#[cfg(test)]
mod tests {
    use super::*;

    fn list(values: &[i32]) -> Option<Box<ListNode>> {
        values
            .iter()
            .rev()
            .fold(None, |next, &val| Some(Box::new(ListNode { val, next })))
    }

    fn list_values(head: &Option<Box<ListNode>>) -> Vec<i32> {
        let mut values = Vec::new();
        let mut current = head.as_deref();
        while let Some(node) = current {
            values.push(node.val);
            current = node.next.as_deref();
        }
        values
    }

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(
            list_values(&Solution::swap_nodes(list(&[1, 2, 3, 4, 5]), 2)),
            vec![1, 4, 3, 2, 5]
        );
        assert_eq!(
            list_values(&Solution::swap_nodes(
                list(&[7, 9, 6, 6, 7, 8, 3, 0, 9, 5]),
                5
            )),
            vec![7, 9, 6, 6, 8, 7, 3, 0, 9, 5]
        );
    }
}
